src/run.py

```diff
@@ -0,0 +1,138 @@
+            f"read file: {metadata_yaml}, content: {metadata!r}"
+        )
+
+    return metadata
+
+
+def _get_key(metadata: dict, key: str) -> str:
+    """Check and return the key value from the metadata.
+
+    Raises InputError if key does not exists, is empty or not a string.
+
+    Args:
+        metadata: The metadata to retrieve the key from.
+
+    Returns:
+        The value of the key.
+
+    """
+    if key not in metadata:
+        raise InputError(f"{key!r} not defined in metadata.yaml, {metadata=!r}")
+    if not isinstance(docs_value := metadata[key], str):
+        raise InputError(f"{key!r} is not a string in metadata.yaml, {metadata=!r}")
+    if not docs_value:
+        raise InputError(f"{key!r} is empty in metadata.yaml, {metadata=!r}")
+    return docs_value
+
+
+def _read_docs_index(base_path: Path) -> str:
+    """Read the content of the index file.
+
+    Raises InputError if the file does not exist.
+
+    Args:
+        base_path: The starting path to look for the index content.
+
+    Returns:
+        The content of the index file.
+
+    """
+    if not (docs_folder := base_path / "docs").is_dir():
+        raise InputError(
+            f"Could not find directory '{docs_folder}' which is where documentation is expected "
+            "to be stored"
+        )
+    if not (index_file := docs_folder / "index.md").is_file():
+        raise InputError(
+            f"Could not find file '{index_file}' which is where the documentation index file is "
+            "expected to be stored"
+        )
+
+    return index_file.read_text()
+
+
+def retrieve_or_create_index(
+    create_if_not_exists: bool, base_path: Path, server_client: Discourse
+) -> Page:
+    """Retrieve the index page defined in the metadata.yaml file or create it if it doesn't exist.
+
+    Raises InputError if create_if_not_exists is False and the docs key is not defined in the
+    metadata.yaml file. Raises ServerError if interactions with the documentation server
+    (retrieving or creating the index page) occurs.
+
+    Args:
+        create_if_not_exists: Whether to create the index page if it does not exist.
+        base_path: The base path to look for the metadata.yaml file in.
+        server_client: A client to the documentation server.
+
+    Returns:
+        The index page.
+
+    """
+    metadata = _get_metadata(base_path=base_path)
+
+    docs_key = "docs"
+    if docs_key not in metadata and create_if_not_exists:
+        name_value = _get_key(metadata=metadata, key="name")
+        content = _read_docs_index(base_path=base_path)
+
+        try:
+            index_url = server_client.create_topic(
+                title=f"{name_value.replace('-', ' ').title()} Documentation Overview",
+                content=content,
+            )
+        except DiscourseError as exc:
+            raise ServerError("Index page creation failed") from exc
+    elif docs_key not in metadata and not create_if_not_exists:
```

Comment:
I tend to always put the error case first in my logic, and then follow the "happy case"
It'll allow to factor a bit the code by using "_get_key" once.
Just a suggestion

Reply:
Yes putting the elif as a separate condition first is better, that should be changed in the next commit. I don't follow being able to reduce the `_get_key` call, the two calls in this function are for different keys, one for the `docs` if that is defined in the metadata indicating that there is an existing documentation post and the other for the `name` key if the `docs` key is not in the metadata which indicates that there is no existing documentation post, in which case it is created. Or am I misunderstanding something?

Reply:
No misundertanding, I didn't read properly the code. Fine if we agree on the `elif`
If you don't mind I'll close the comment once the change is introduced ?

Reply:
It should already be done